<div style="padding: 28px 32px; border: 1px solid #bfdbfe; border-left: 8px solid #2563eb; border-radius: 12px; background: linear-gradient(135deg, #eff6ff 0%, #f8fafc 100%);">

<h1 style="margin: 0; color: #1e3a8a;">Machine Learning Zoomcamp 2026</h1>
<p style="margin: 8px 0 18px; font-size: 1.2em; color: #334155;"><strong>Homework 1</strong> · Introduction to Machine Learning</p>

| Student | Cohort | Dataset | Environment |
| :--- | :--- | :--- | :--- |
| Carlos Ibarra | 2026 | 2026 Car Fuel Efficiency | Python 3.11+ · Pandas 3.0.6 · NumPy |

</div>

### Learning path

| Step | Focus |
| :---: | :--- |
| 01 | Environment setup and library validation |
| 02 | Data ingestion and shape inspection |
| 03 | Categorical feature exploration |
| 04 | Missing value audit |
| 05 | Conditional filtering and aggregation |
| 06 | Imputation impact analysis |
| 07 | Normal equation with NumPy |

<div style="padding: 18px 22px; border-left: 6px solid #2563eb; border-radius: 8px; background: #eff6ff;">

## Q1 · Pandas version

**Question.** What version of Pandas is installed in the environment?

</div>

In [1]:
import pandas as pd
pd.__version__

'3.0.6'

<div style="padding: 14px 18px; border-left: 5px solid #16a34a; border-radius: 8px; background: #f0fdf4;">

**Result · Q1**
The installed Pandas version is **`3.0.6`**.

</div>

<div style="padding: 18px 22px; border-left: 6px solid #2563eb; border-radius: 8px; background: #eff6ff;">

## Q2 · Records count

**Question.** How many records are available in the dataset?

</div>

| Option | Records |
| :---: | ---: |
| A | 5,000 |
| B | 9,000 |
| C | 10,000 |
| D | 15,000 |

In [2]:
# Load the dataset
data_path = 'data/car_fuel_efficiency_2026.csv'
df = pd.read_csv(data_path)
df.shape

(10000, 11)

<div style="padding: 14px 18px; border-left: 5px solid #16a34a; border-radius: 8px; background: #f0fdf4;">

**Result · Q2**
The dataset contains **`10,000`** records.

</div>

<div style="padding: 18px 22px; border-left: 6px solid #2563eb; border-radius: 8px; background: #eff6ff;">

## Q3 · Fuel types

**Question.** How many fuel types are represented in the dataset?

</div>

| Option | Number of fuel types |
| :---: | ---: |
| A | 1 |
| B | 2 |
| C | 3 |
| D | 4 |

In [3]:
for column in df.columns:
    print(f"- {column}")

- model_year
- origin
- fuel_type
- drivetrain
- num_doors
- engine_displacement
- num_cylinders
- horsepower
- vehicle_weight
- acceleration
- fuel_efficiency_mpg


In [4]:
fuel_type_counts = df['fuel_type'].value_counts(dropna=False)
fuel_type_counts

fuel_type
Gasoline    6913
Diesel      2053
Hybrid      1034
Name: count, dtype: int64

<div style="padding: 14px 18px; border-left: 5px solid #16a34a; border-radius: 8px; background: #f0fdf4;">

**Result · Q3**
There are **`3`** unique fuel types: `Gasoline` (6,913), `Diesel` (2,053), and `Hybrid` (1,034).

</div>

<div style="padding: 18px 22px; border-left: 6px solid #2563eb; border-radius: 8px; background: #eff6ff;">

## Q4 · Missing values

**Question.** How many columns contain missing values?

</div>

| Option | Columns with missing values |
| :---: | ---: |
| A | 0 |
| B | 1 |
| C | 2 |
| D | 3 |
| E | 4 |

In [5]:
# Inspect missing values per column
df.isnull().sum()

model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors                0
engine_displacement      0
num_cylinders            0
horsepower             877
vehicle_weight           0
acceleration           264
fuel_efficiency_mpg      0
dtype: int64

<div style="padding: 14px 18px; border-left: 5px solid #16a34a; border-radius: 8px; background: #f0fdf4;">

**Result · Q4**
Exactly **`2`** columns have missing values: `horsepower` (877 NaNs) and `acceleration` (264 NaNs).

</div>

<div style="padding: 18px 22px; border-left: 6px solid #2563eb; border-radius: 8px; background: #eff6ff;">

## Q5 · Maximum fuel efficiency

**Question.** What is the maximum fuel efficiency of cars manufactured in Asia?

</div>

| Option | Fuel efficiency (mpg) |
| :---: | ---: |
| A | 21.2 |
| B | 31.2 |
| C | 41.2 |
| D | 51.2 |

In [6]:
# Filter cars from Asia
asia_cars = df[df["origin"] == "Asia"]

# Calculate the maximum fuel efficiency
asia_cars["fuel_efficiency_mpg"].max()

np.float64(41.2)

<div style="padding: 14px 18px; border-left: 5px solid #16a34a; border-radius: 8px; background: #f0fdf4;">

**Result · Q5**
The maximum fuel efficiency for cars manufactured in Asia is **`41.2` mpg**.

</div>

<div style="padding: 18px 22px; border-left: 6px solid #2563eb; border-radius: 8px; background: #eff6ff;">

## Q6 · Median value of horsepower

**Question.** Does mode-based imputation change the median of `horsepower`?

</div>

| Step | Operation |
| :---: | :--- |
| 1 | Calculate the original median. |
| 2 | Find the most frequent value. |
| 3 | Fill missing values with that mode. |
| 4 | Calculate the median again. |

| Option | Outcome |
| :---: | :--- |
| A | Yes, it increased |
| B | Yes, it decreased |
| C | No |

In [7]:
# Calculate the median before imputation
median_before = df["horsepower"].median()

# Find the most frequent horsepower value
mode_value = df["horsepower"].mode()[0]

# Fill missing values without changing the original DataFrame
horsepower_filled = df["horsepower"].fillna(mode_value)

# Calculate the median after imputation
median_after = horsepower_filled.median()

pd.Series(
    {
        "Median before imputation": median_before,
        "Mode used for imputation": mode_value,
        "Median after imputation": median_after,
    }
)

Median before imputation    254.0
Mode used for imputation    252.0
Median after imputation     252.0
dtype: float64

<div style="padding: 14px 18px; border-left: 5px solid #16a34a; border-radius: 8px; background: #f0fdf4;">

**Result · Q6**
**`Yes, it decreased`** — from **`254.0`** to **`252.0`**.

<small>Imputing 877 missing values with the mode (252.0) added mass below the original median.</small>

</div>

<div style="padding: 18px 22px; border-left: 6px solid #2563eb; border-radius: 8px; background: #eff6ff;">

## Q7 · Sum of regression weights

**Question.** What is the sum of the weights returned by the normal equation?

</div>

| Step | Linear algebra operation |
| :---: | :--- |
| 1 | Select the first 7 cars from Asia. |
| 2 | Build `X` from `vehicle_weight` and `model_year`. |
| 3 | Compute `XTX = X.T @ X` and invert it. |
| 4 | Calculate `w = inv(XTX) @ X.T @ y`. |
| 5 | Return `w.sum()`. |

| Option | Sum of weights |
| :---: | ---: |
| A | 0.0369 |
| B | 0.369 |
| C | 3.69 |
| D | 36.9 |

In [8]:
import numpy as np

# Select the first seven cars from Asia and the required features
X = df.loc[df["origin"] == "Asia", ["vehicle_weight", "model_year"]].iloc[:7].to_numpy()

# Calculate the normal equation components
XTX = X.T @ X
inv_XTX = np.linalg.inv(XTX)

# Define the target vector and calculate the weights
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
w = inv_XTX @ X.T @ y

w.sum()

np.float64(0.36919696904925203)

<div style="padding: 14px 18px; border-left: 5px solid #16a34a; border-radius: 8px; background: #f0fdf4;">

**Result · Q7**
The sum of all elements in vector $w$ is approximately **`0.369`** (`0.3691969...`).

</div>

<div style="padding: 22px 26px; border: 1px solid #bfdbfe; border-top: 7px solid #2563eb; border-radius: 12px; background: #f8fafc;">

## Homework 1 · Results dashboard

| Question | Focus | Final answer |
| :---: | :--- | :--- |
| **Q1** | Pandas version | **`3.0.6`** |
| **Q2** | Records count | **`10,000`** |
| **Q3** | Fuel types | **`3`** |
| **Q4** | Missing-value columns | **`2`** |
| **Q5** | Maximum fuel efficiency | **`41.2` mpg** |
| **Q6** | Median after imputation | **Yes, it decreased** |
| **Q7** | Sum of regression weights | **`0.369`** |

</div>

<div style="margin-top: 16px; padding: 16px 20px; border-left: 5px solid #f59e0b; border-radius: 8px; background: #fffbeb;">

### Key takeaway

Imputation changes a variable's distribution. In Q6, filling missing `horsepower` values with the mode moved the median from **254.0** to **252.0**. Validate these effects before using processed features in a machine learning model.

</div>